### PR TITLE
Moved out statefulset status update for ETCD to separate controller.

### PR DESCRIPTION
**Pull request by abdasgupta**
**What this PR does / why we need it**:
Moved parts of reconciliation for ETCD status that deals with Statefulset to a separate controller.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/etcd-druid/issues/106

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
Moved parts of reconciliation for ETCD status that deals with Statefulset to a separate controller.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Any update on this? We have hit it again last week.

**Comment by amshuman-kr** [comment]
We are planning to merge this week of early next week but wait for release with other changes (like the backup folder structure changes as proposed in #132). But if required these can be release separately.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
While testing integration with gardener, I found the following issues. Can you please address them?

- There is an RBAC problem with the `ClusterRole` deployed by gardener for `etcd-druid`. I think this is an issue introduced with one of the recently merged PR. I think something like the following needs to be added to the `ClusterRole` created in gardener [here](https://github.com/gardener/gardener/blob/9df08ae3ae865f96db4161eb4ad01ced73107f3e/pkg/operation/botanist/component/etcd/bootstrap.go#L145,).
```go
{
	APIGroups: []string{coordinationv1.GroupName},
	Resources: []string{"leases"},
	ResourceNames: []string{"druid-leader-election"},
	Verbs:     []string{"get", "update", "patch", "create"},
},
```
- The [vendored](https://github.com/gardener/etcd-druid/blob/a1db8795351b44bc6f57e1e1e5daf3a2e0edde10/vendor/github.com/gardener/gardener/pkg/utils/kubernetes/health/health.go#L131-L149) `CheckStatefulSet` func checks if the `StatefulSet` `spec.replicas` matches the `status.readyReplicas`. But this doesn't cover the case where some operator manually scales down the etcd `StatefulSet` without hibernating the shoot cluster where it still makes sense to expose that the etcd is not ready in the `Etcd` resource status. I think, we should make a copy of the `CheckStatefulSet` func to compare the `Etcd` resource `spec.replicas` with the `StatefulSet` `status.readyReplicas`.
- Crashing the etcd pod (by editing the `StatefulSet` spec to add some junk into the command section) resulted in the `StatefulSet` status being updated properly (no `readyReplicas`) but didn't update the `Etcd` status as expected. It still had `ready: true, readyReplicas: 1`.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
The tests still fail.

I scaled the ETCD statefulsets down and the `Etcd` resources are still marked `Ready`.

```
$ k -n <shoot namespace> get sts,etcd
NAME                           READY   AGE
statefulset.apps/etcd-events   0/0     143m
statefulset.apps/etcd-main     0/0     144m
statefulset.apps/loki          1/1     144m
statefulset.apps/prometheus    1/1     136m

NAME                                    READY   AGE
etcd.druid.gardener.cloud/etcd-events   true    143m
etcd.druid.gardener.cloud/etcd-main     true    144m
```

**Comment by amshuman-kr** [comment]
@abdasgupta If it helps, the seed cluster I am using is on Kubernetes `v1.18.16`.

**Review by amshuman-kr — COMMENTED** [review]
@abdasgupta I tested the PR again and the `Etcd` resource status update is now working based on `StatefulSet` status update. Have you found out why the unit test was working but not the gardener integration test? Also, did you make any changes in the informer/watch connection for the statefulset?

The testing of the PR looks fine. If you can confirm the change the was required and if the change is also included in the unit-tests, then we can merge this PR.

**Comment by abdasgupta** [comment]
@amshuman-kr I made a very small change in the controller manager for custodian controller to watch for statefulset mainly instead of ETCD resource.

File: controllers/etcd_custodian_controller.go
Comment (amshuman-kr) [inline controllers/etcd_custodian_controller.go:162]: Like I mentioned in out-of-band discussion, it would have been better to do this using `For` on `Etcd` and `Watches` on `StatefulSet`. But I am OK if you want to do that in a different PR later. @abdasgupta Please create an issue for this so that we do not forget.

Approving for now.

**Review by amshuman-kr — APPROVED** [review]
One nit-pick still below as discussed out-of-band but can be merged provided we track the nit-pick in another issue and address it later.

**Review by shreyas-s-rao — APPROVED** [review]
@abdasgupta thanks for making the requested changes from the out-of-band discussion. LGTM

**Pull request merged**
